app/domain/authentication/authn_azure/application_identity.rb

```diff
@@ -0,0 +1,132 @@
+module Authentication
+  module AuthnAzure
+
+    Log = LogMessages::Authentication
+    Err = Errors::Authentication::AuthnAzure
+    # Possible Errors Raised: ConstraintNotSupported, MissingAnnotationConstraint, IllegalConstraintCombinations
+
+    # This class defines an application identity of a given Conjur host.
+    # The constructor initializes an ApplicationIdentity object and validates that
+    # it is configured correctly.
+    # The difference between the validation in this constructor and
+    # the validation in ValidateApplicationIdentity is that here we validate that
+    # the application identity is configured correctly, and thus is a valid application
+    # identity. In ValidateApplicationIdentity we validate that the defined application
+    # identity is actually the one we expect from Azure. This validation is done through
+    # a combination of comparisons that are performed based on the type of assigned
+    # identity provided.
+    # For example, an application identity `some-namespace/blah/some-value` is not
+    # valid and will fail the validation here. However, an application identity
+    # `some-namespace/service-account/some-value` is a valid application identity
+    # and will pass the validation here.
+    # If the Azure-related annotations in the host defined in Conjur has Azure-specific
+    # annotations that do not match the supplied JWT token, then it will fail the
+    # validation of ValidateApplicationIdentity.
+    class ApplicationIdentity
+
+      def initialize(role_annotations:, service_id:)
+        @role_annotations = role_annotations
+        @service_id       = service_id
+
+        validate
+      end
+
+      def constraints
+        @constraints ||= {
+            subscription_id:          constraint_value("subscription-id"),
+            resource_group:           constraint_value("resource-group"),
+            user_assigned_identity:   constraint_value("user-assigned-identity"),
+            system_assigned_identity: constraint_value("system-assigned-identity"),
+        }.compact
+      end
+
+      private
+
+      # validate that the application identity is defined correctly
+      def validate
+        # check if annotations present are part of the permitted annotations for authn-azure
+        validate_permitted_annotations
+
+        # validate that a constraint exists on the required annotations
+        validate_required_constraints_exist
+        validate_constraint_combinations
+      end
+
+      def validate_permitted_annotations
+        validate_prefixed_permitted_annotations("authn-azure/")
+        validate_prefixed_permitted_annotations("authn-azure/#{@service_id}/")
+      end
+
+      # check if annotations with prefix is part of the permitted list
+      def validate_prefixed_permitted_annotations prefix
+        Rails.logger.debug(Log::ValidatingAnnotationsWithPrefix.new(prefix))
```

Comment:
`Rails.logger` is an external dependency and should be injected.  One of the main benefits of having all of this auth code outside of rails in app/domain is that it has no dependency on rails.  This breaks that.

Reply:
So you would change this class to a command class called `init_application_identity`? or something like that.